reactive/flannel.py

```diff
@@ -1,254 +1,190 @@
-    a mandatory pre-requisite before we can progress '''
-    # this is likely to run first, save the TLS credentials
-    if is_state('etcd.tls.available'):
-        cert_path = '/etc/ssl/flannel'
-        etcd.save_client_credentials('{}/client-key.pem'.format(cert_path),
-                                     '{}/client-cert.pem'.format(cert_path),
-                                     '{}/client-ca.pem'.format(cert_path))
-
-    set_state('flannel.etcd.credentials.placed')
-
-
-@when_not('flannel.installed')
-def install_flannel():
-    ''' Unpack and install the binary release of flannel '''
-    hookenv.status_set('maintenance', 'Installing flannel.')
-    flannel_package = hookenv.resource_get('flannel')
-
-    if not flannel_package:
-        hookenv.status_set('blocked', 'Missing flannel resource.')
+    if not archive:
+        message = 'Missing flannel resource.'
+        log(message)
+        status_set('blocked', message)
         return
-
-    # Handle null resource publication, we check if its filesize < 1mb
-    filesize = os.stat(flannel_package).st_size
+    filesize = os.stat(archive).st_size
     if filesize < 1000000:
-        hookenv.status_set('blocked', 'Incomplete flannel resource.')
+        message = 'Incomplete flannel resource'
+        log(message)
+        status_set('blocked', message)
         return
-
-    charm_dir = hookenv.charm_dir()
-
-    # Unpack and install the flannel resource
-    _unpack_resource(flannel_package, '{0}/files/flannel'.format(charm_dir))
-    _install_files('./files/flannel/flanneld', '/usr/local/bin/flanneld')
-
-    set_state('flannel.installed')
-
-
-@when('flannel.installed', 'flannel.etcd.credentials.placed', 'etcd.available')
-@when_not('flannel.sdn.configured')
-def render_flannel_config(etcd):
-    ''' Consume the flannel information, and render the config files to
-    initialize flannel installation '''
-    hookenv.status_set('maintenance', 'Rendering systemd files.')
-    context = _build_context(etcd)
+    status_set('maintenance', 'Unpacking flannel resource.')
+    charm_dir = os.getenv('CHARM_DIR')
+    unpack_path = os.path.join(charm_dir, 'files', 'flannel')
+    os.makedirs(unpack_path, exist_ok=True)
+    cmd = ['tar', 'xfz', archive, '-C', unpack_path]
+    log(cmd)
+    check_call(cmd)
+    apps = [
+        {'name': 'flanneld', 'path': '/usr/local/bin'},
+        {'name': 'etcdctl', 'path': '/usr/local/bin'},
+        {'name': 'flannel', 'path': '/opt/cni/bin'},
+        {'name': 'bridge', 'path': '/opt/cni/bin'},
+        {'name': 'host-local', 'path': '/opt/cni/bin'}
+    ]
+    for app in apps:
+        unpacked = os.path.join(unpack_path, app['name'])
+        app_path = os.path.join(app['path'], app['name'])
+        install = ['install', '-v', '-D', unpacked, app_path]
+        check_call(install)
+    set_state('flannel.binaries.installed')
+
+
+@when_not('flannel.cni.configured', 'cni.is-worker')
+def configure_cni():
```

Comment:
I guess maybe this ran before the `cni` relation was handled? In which case, neither `cni.is-worker` or `cni.is-master` would have been set. I think that makes sense.